### PR TITLE
[GraphOptimizer] Move transpose constant folding optimization before merging BatchNorm into Convolution

### PR DESCRIPTION
**Pull request by aturetsk**
*Description*:
An internal net model I'm working on has explicit transpose nodes for convolution weights. This prevents BatchNorm + Conv merge optimization from triggering.
The patch moves transpose constant folding before BatchNorm optimization in order to get rid of these transposes and make BatchNorm + Conv merge happen.

*Testing*:
Added a unit test.

*Documentation*:
N/A

### CODE REVIEW

**Comment by jfix71** [comment]
I tried running the unit test you added here without swapping the order of `optimizeTranspose` and `optimizeBatchNorm` and it still passed. Can you make sure the test exposes and tests the problem? I'm also a bit unsure what is preventing the Conv-BatchNorm merging from occurring. Can you explain a little more?

**Comment by aturetsk** [comment]
@jfix71,
Fixed the test.

The problem is that optimizeBatchNorm expects convolution weights and bias to be Variables. If it's not so (for example, if there is a transpose node between convolution and weight variable), the optimization is not triggered. So if we run transpose constant folding before optimizeBatchNorm this issue can be avoided.

File: tests/unittests/graphOptzTest.cpp [hunk 1]
```diff
@@ -161,6 +161,32 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvFP16) {
...
+  auto *ret = F_->createSave(ctx_, "ret", BN);
+
+  // Initialize to ensure that constant tensors are not optimized out.
+  ctx_.allocate(filter)->getHandle().randomize(-1.0, 1.0, mod_.getPRNG());
+  ctx_.allocate(bias)->getHandle().randomize(-1.0, 1.0, mod_.getPRNG());
+  ::glow::convertPlaceholdersToConstants(F_, ctx_, {});
+
+  EXPECT_EQ(F_->getNodes().size(), 4);
+  ::glow::optimize(F_, CompilationMode::Infer);
+  EXPECT_EQ(F_->getNodes().size(), 2);
```
Comment (jfix71) [inline tests/unittests/graphOptzTest.cpp]: nit: Can you add a check that there are no BatchNorm nodes left? Can be simple using `countNodeKind`.

**Comment by aturetsk** [comment]
@jfix71, thanks, done!

File: tests/unittests/graphOptzTest.cpp [hunk 2]
```diff
@@ -161,6 +161,33 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvFP16) {
   EXPECT_TRUE(llvm::isa<SaveNode>(save));
 }
 
+/// Check that transpose constant folding is done before BatchNorm optimization,
+/// which allows to merge BatchNorm into Convolution with transposed weights.
+TEST_F(GraphOptz, optimizeBatchNormAfterConvWithTransposedWeights) {
+  Node *input =
+      mod_.createPlaceholder(ElemKind::FloatTy, {1, 10, 20, 3}, "A", false);
+  auto *filter =
+      mod_.createPlaceholder(ElemKind::FloatTy, {16, 3, 5, 5}, "filter", true);
```
Comment (rdzhabarov) [inline tests/unittests/graphOptzTest.cpp]: not sure why these are trainable :) any reason?

File: tests/unittests/graphOptzTest.cpp [hunk 3]
```diff
@@ -161,6 +161,33 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvFP16) {
...
+  auto *CV = F_->createConv("conv", input, TN, bias,
+                            mod_.uniqueType(ElemKind::FloatTy, {1, 10, 20, 16}),
+                            5, 1, 2, 1);
+  auto *BN = F_->createBatchNormalization(ctx_, "batch", CV, 3, 0.0001, 0.9);
+  auto *ret = F_->createSave(ctx_, "ret", BN);
+
+  // Initialize to ensure that constant tensors are not optimized out.
+  ctx_.allocate(filter)->getHandle().randomize(-1.0, 1.0, mod_.getPRNG());
+  ctx_.allocate(bias)->getHandle().randomize(-1.0, 1.0, mod_.getPRNG());
+  ::glow::convertPlaceholdersToConstants(F_, ctx_, {});
```
Comment (rdzhabarov) [inline tests/unittests/graphOptzTest.cpp]: typically input is not converted to constant.

File: tests/unittests/graphOptzTest.cpp (hunk 2 above)
Comment (aturetsk) [reply tests/unittests/graphOptzTest.cpp]: No particular reason, the optimization should work with both.
Is there any reason to use un-trainable placeholder vs trainable in tests? Both kinds seem to be used randomly in tests...

Comment (rdzhabarov) [reply tests/unittests/graphOptzTest.cpp]: it's just when I see that var is trainable I start to think whether there are any reasons behind that, or whether it has to be trainable.

**Comment by aturetsk** [comment]
@rdzhabarov, made variables untrainable and included input in convertPlaceholdersToConstants exclusion list.

**Comment by rdzhabarov** [comment]
it looks like "Merge branch 'master' into cadence/turetski/improve-batch-norm-opt" does not really add anything to commits here, so I'm going to merge this.

**Comment by rdzhabarov** [comment]
thank you @aturetsk!
